File: filecrypt.rb
#!/usr/bin/env ruby
# filecrypt.rb
# encoding: UTF-8

require 'openssl'
require 'base64'
require 'fileutils'

# ANSI colors
COLORS = {
  green: "\e[92m",
  red: "\e[91m",
  yellow: "\e[93m",
  reset: "\e[0m"
}

def colorize(text, color)
  "#{COLORS[color]}#{text}#{COLORS[:reset]}"
end

SALT_LEN = 16
NONCE_LEN = 12
TAG_LEN = 16
KEY_LEN = 32
ITERATIONS = 100000
CHUNK_SIZE = 64 * 1024

def derive_key(password, salt)
  OpenSSL::PKCS5.pbkdf2_hmac(password, salt, ITERATIONS, KEY_LEN, 'sha256')
end

def encrypt_file(input_path, output_path, password)
  salt = OpenSSL::Random.random_bytes(SALT_LEN)
  nonce = OpenSSL::Random.random_bytes(NONCE_LEN)
  key = derive_key(password, salt)

  cipher = OpenSSL::Cipher.new('aes-256-gcm')
  cipher.encrypt
  cipher.key = key
  cipher.iv = nonce

  plaintext = File.binread(input_path)
  ciphertext = cipher.update(plaintext) + cipher.final
  tag = cipher.auth_tag

  File.open(output_path, 'wb') do |f|
    f.write(salt)
    f.write(nonce)
    f.write(ciphertext)
    f.write(tag)
  end
  puts colorize("File encrypted successfully: #{output_path}", :green)
end

def decrypt_file(input_path, output_path, password)
  data = File.binread(input_path)
  if data.bytesize < SALT_LEN + NONCE_LEN + TAG_LEN
    raise "Invalid file format"
  end

  salt = data.byteslice(0, SALT_LEN)
  nonce = data.byteslice(SALT_LEN, NONCE_LEN)
  ciphertext = data.byteslice(SALT_LEN + NONCE_LEN, data.bytesize - SALT_LEN - NONCE_LEN - TAG_LEN)
  tag = data.byteslice(-TAG_LEN, TAG_LEN)

  key = derive_key(password, salt)
  decipher = OpenSSL::Cipher.new('aes-256-gcm')
  decipher.decrypt
  decipher.key = key
  decipher.iv = nonce
  decipher.auth_tag = tag

  plaintext = decipher.update(ciphertext) + decipher.final
  File.binwrite(output_path, plaintext)
  puts colorize("File decrypted successfully: #{output_path}", :green)
end

def main
  if ARGV.size < 4
    puts colorize("Usage: ruby filecrypt.rb encrypt|decrypt <input> <output> <password>", :yellow)
    exit 1
  end

  mode, input, output, password = ARGV[0], ARGV[1], ARGV[2], ARGV[3]

  begin
    if mode == 'encrypt'
      encrypt_file(input, output, password)
    elsif mode == 'decrypt'
      decrypt_file(input, output, password)
    else
      puts colorize("Invalid mode. Use encrypt or decrypt.", :red)
      exit 1
    end
  rescue => e
    puts colorize("Error: #{e.message}", :red)
    exit 1
  end
end

if __FILE__ == $0
  main
end

File: filecrypt.js
// filecrypt.js
#!/usr/bin/env node
'use strict';

const crypto = require('crypto');
const fs = require('fs');

// ANSI colors
const COLORS = {
    green: '\x1b[92m',
    red: '\x1b[91m',
    yellow: '\x1b[93m',
    reset: '\x1b[0m'
};

function colorize(text, color) {
    return COLORS[color] + text + COLORS.reset;
}

const SALT_LEN = 16;
const NONCE_LEN = 12;
const TAG_LEN = 16;
const KEY_LEN = 32;
const ITERATIONS = 100000;
const CHUNK_SIZE = 64 * 1024;

function deriveKey(password, salt) {
    return crypto.pbkdf2Sync(password, salt, ITERATIONS, KEY_LEN, 'sha256');
}

function encryptFile(inputPath, outputPath, password) {
    const salt = crypto.randomBytes(SALT_LEN);
    const nonce = crypto.randomBytes(NONCE_LEN);
    const key = deriveKey(password, salt);

    const plaintext = fs.readFileSync(inputPath);
    const cipher = crypto.createCipheriv('aes-256-gcm', key, nonce);
    const encrypted = Buffer.concat([cipher.update(plaintext), cipher.final()]);
    const tag = cipher.getAuthTag();

    const out = Buffer.concat([salt, nonce, encrypted, tag]);
    fs.writeFileSync(outputPath, out);
    console.log(colorize(`File encrypted successfully: ${outputPath}`, 'green'));
}

function decryptFile(inputPath, outputPath, password) {
    const data = fs.readFileSync(inputPath);
    const salt = data.slice(0, SALT_LEN);
    const nonce = data.slice(SALT_LEN, SALT_LEN + NONCE_LEN);
    const encrypted = data.slice(SALT_LEN + NONCE_LEN, -TAG_LEN);
    const tag = data.slice(-TAG_LEN);
    const key = deriveKey(password, salt);

    const decipher = crypto.createDecipheriv('aes-256-gcm', key, nonce);
    decipher.setAuthTag(tag);
    const plaintext = Buffer.concat([decipher.update(encrypted), decipher.final()]);
    fs.writeFileSync(outputPath, plaintext);
    console.log(colorize(`File decrypted successfully: ${outputPath}`, 'green'));
}

function main() {
    const args = process.argv.slice(2);
    if (args.length < 4) {
        console.log(colorize('Usage: node filecrypt.js encrypt|decrypt <input> <output> <password>', 'yellow'));
        process.exit(1);
    }
    const mode = args[0];
    const input = args[1];
    const output = args[2];
    const password = args[3];

    try {
        if (mode === 'encrypt') encryptFile(input, output, password);
        else if (mode === 'decrypt') decryptFile(input, output, password);
        else {
            console.log(colorize('Invalid mode. Use encrypt or decrypt.', 'red'));
            process.exit(1);
        }
    } catch (err) {
        console.log(colorize(`Error: ${err.message}`, 'red'));
        process.exit(1);
    }
}

main();
